### PR TITLE
Add randomizedDelaySec, persistent options to nix-gc and auto-upgrade

### PR DESCRIPTION
**Pull request by tex**
###### Motivation for this change

I am complaining about auto-upgrade (and nix-gc) not correctly working in laptop environment where laptop is turned on and off unpredictable causing those timers never expire and those services never executed.

Here is my proposal to fix it. Please advice, comment. Since I am changing existing API I guess it is not for direct delivery (unless it is), but I think that current situation is making those services unusable on laptops. Since it is auto upgrade that could lead to security issues as user might expect getting automatic updates while he is really not.

#15689
#75861

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by xaverdh** [comment]
It should be possible to keep both apis in parallel, at least for now?

**Comment by tex** [comment]
Good idea. I changed it, so if user already has set it's own dates then it will use them. But defaults are changes to work in laptop environment (and that doesn't break servers). Also updated comments a bit to indicate that dates accepts fixed date/time or intervals. And the Persistent is important to make intervals works on laptops.

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
I like the idea. I made nix-gc persistent on the side for the same reason, but it's nicer to have a proper fix. I will test both modules on my infrastructure this week.

Your commit doesn't follow CONTRIBUTING.md. Ideally, you should split it into two, namely
- nixos/nix-gc: add persistent and randomizeDelaySec options
- nixos/auto-upgrade: add persistent option

File: nixos/modules/services/misc/nix-gc.nix [hunk 1]
```diff
@@ -21,13 +21,45 @@ in
       };
 
       dates = mkOption {
-        default = "03:15";
         type = types.str;
+        default = "weekly";
+        example = "3:15";
         description = ''
-          Specification (in the format described by
+          How often or when upgrade occurs. For most desktop and server systems
```
Comment (lourkeur) [inline nixos/modules/services/misc/nix-gc.nix:28]:
```suggestion
          How often or when garbage collection is performed. For most desktop and server systems
```

File: nixos/modules/tasks/auto-upgrade.nix [hunk 2]
```diff
@@ -63,13 +63,16 @@ in {
...
 
       dates = mkOption {
-        default = "04:40";
         type = types.str;
+        default = "daily";
+        example = "04:40";
         description = ''
-          Specification (in the format described by
+          How often or when upgrade occurs. For most desktop and server systems
+          a sufficient trimming frequency is once a day.
```
Comment (lourkeur) [inline nixos/modules/tasks/auto-upgrade.nix:71]:
```suggestion
          a sufficient upgrade frequency is once a day.
```

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
Sorry for the traffic, I fumbled posting those comments apparently

**Comment by tex** [comment]
I see you config:
```
  nix.gc.automatic = true;
  nix.gc.options = "--delete-older-than 1w";
  nix.gc.dates = "Sat 03:15";
  systemd.timers.nix-gc.timerConfig.Persistent = true;
```
Oh, I see I could set Persistent directly in my configuration.nix! :) I didn't know it can be so easy :)

But fix it in core is definitely better, at least for users and discoverability of the configuration.

Will split it to two merge requests and apply your suggested changes.

**Comment by tex** [comment]
Created two merge requests according to your suggestion. Closing this one.

**Pull request closed without merging**